common:
    lidar_topic:  "/velodyne_points"
    imu_topic:  "/imu/data"
    point_filter_num: 4
    sweep_cut_num: 1
    time_sync_en: false         # ONLY turn on when external time synchronization is really not possible
    gravity_acc: [ 0.0, 0.0, 9.79744]
    
lidar_parameter:
    lidar_type: 2                # 1 for Livox serials LiDAR, 2 for Velodyne LiDAR, 3 for ouster LiDAR, 
    N_SCANS: 32
    SCAN_RATE: 10                # only need to be set for velodyne, unit: Hz,
    time_unit: 0            # the unit of time/t field in the PointCloud2 rostopic: 0-second, 1-milisecond, 2-microsecond, 3-nanosecond.
    blind: 2
    fov_degree:    180
    det_range:     100.0

imu_parameter:
    acc_cov: 0.1
    gyr_cov: 0.1
    b_acc_cov: 0.0001
    b_gyr_cov: 0.0001
    time_diff_enable: false

extrinsic_parameter:
    extrinsic_enable:  false      # true: enable the online estimation of IMU-LiDAR extrinsic,
    extrinsic_t: [ -0.5, 1.4, 1.5]
    extrinsic_R: [ 1, 0, 0, 
                   0, 1, 0, 
                   0, 0, 1]

publish:
    path_en:  false
    scan_publish_en:  true       # false: close all the point cloud output
    dense_publish_en: true       # false: low down the points number in a global-frame point clouds scan.
    scan_bodyframe_pub_en: true  # true: output the point cloud scans in IMU-body-frame

pcd_save:
    pcd_save_en: true
    interval: -1                 # how many LiDAR frames saved in each pcd file; 
                                 # -1 : all frames will be saved in ONE pcd file, may lead to memory crash when having too much frames.

odometry_options:
    voxel_size: 0.5 # The voxel size for the grid sampling of the new frame (before keypoints extraction)
    sample_voxel_size: 1.5 # The size of a voxel for the selection of `keypoints` by grid sampling
    max_distance: 2000.0 # The threshold of the distance to suppress voxels from the map
    max_num_points_in_voxel: 20 # The maximum number of points per voxel of the map
    method_system_init: STATIC_INIT # LIO system initialization [MOTION_INIT, STATIC_INIT]
    init_num_frames: 20
    robust_registration: false
    min_distance_points: 0.15
    distance_error_threshold: 100.0 # The motion of the sensor between two frames which is considered erroneous (stops the odometry)
    motion_compensation: CONTINUOUS # The profile of the motion compensation (NONE, CONSTANT_VELOCITY, ITERATIVE, CONTINUOUS)
    initialization: INIT_CONSTANT_VELOCITY # [INIT_CONSTANT_VELOCITY, INIT_IMU]

icp_options:
    size_voxel_map: 1.0 # The voxel size of in the voxel map
    num_iters_icp: 15 # The number of iterations of the ICP
    min_number_neighbors: 20 # The minimum number of neighbor points to define a valid neighborhood
    voxel_neighborhood: 1
    max_number_neighbors: 20
    max_dist_to_plane_ct_icp: 0.3
    threshold_orientation_norm: 0.1 # Threshold on orientation changes (in degrees) for early termination of the ICP
    threshold_translation_norm: 0.01 # Threshold on distance changes (in m) for early termination of the ICP
    debug_print: false
    point_to_plane_with_distortion: true
    distance: CT_POINT_TO_PLANE # Options: [CT_POINT_TO_PLANE, POINT_TO_PLANE]
    num_closest_neighbors: 1
    beta_location_consistency: 1.0
    beta_orientation_consistency: 1.0
    beta_constant_velocity: 1.0 # Trajectory Regularisation Weight to constrain the optimization
    beta_small_velocity: 0.00
    solver: LIDAR # The type of SOLVER used in [LIO, LIDAR]
    min_num_residuals: 600 # The minimum number of residuals for a valid ICP problem
    max_num_residuals: 800 # The maximum number of residuals considered (if more keypoints exist, residuals are randomly sampled)

    loss_function: HUBER # Options: [CAUCHY, STANDARD, HUBER, TOLERANT, TRUNCATED]
    ls_max_num_iters: 5 # The number of steps performed by ceres for each iteration of the ICP
    ls_num_threads: 16 # The number of threads to build and solve the least square system
    ls_sigma: 0.5
    ls_tolerant_min_threshold: 0.05
